File: processors/processor.py
import json
from typing import Dict, Any, List, Optional, Set
from common.logging import setup_logging
import sqlite3

logger = setup_logging()


class DataProcessor:
    def __init__(self, db):
        self.db = db

    def process_tournaments(self, source: str = 'limitless', tournament_ids: Optional[List[str]] = None, force: bool = False) -> Dict[str, Any]:
        """Process raw tournament data into structured tables."""
        
        results = {
            "success": True,
            "tournaments_processed": 0,
            "players_added": 0,
            "teams_added": 0,
            "pokemon_sets_added": 0,
            "matches_added": 0,
            "tournament_standings_added": 0,
            "errors": []
        }

        cursor = self.db.conn.cursor()

        if source == 'limitless':
            query = "SELECT id, details, standings, pairings FROM limitless_api_raw_data"
            params = []
            
            if tournament_ids:
                placeholders = ','.join('?' * len(tournament_ids))
                query += f" WHERE id IN ({placeholders})"
                params = tournament_ids
        else:
            results["success"] = False
            results["errors"].append(f"Unsupported source: {source}")
            return results

        cursor.execute(query, params)
        raw_data_rows = cursor.fetchall()

        logger.info("Found raw tournaments", count=len(raw_data_rows))

        for row in raw_data_rows:
            tournament_id = row[0]
            details_json = row[1]
            standings_json = row[2]
            pairings_json = row[3]

            try:
                if not details_json or not standings_json or not pairings_json:
                    logger.warning("Incomplete data for tournament", id=tournament_id)
                    continue

                details = json.loads(details_json)
                standings = json.loads(standings_json)
                pairings = json.loads(pairings_json)

                if not force:
                    if self._is_processed(tournament_id):
                        logger.info("Tournament already processed, skipping", id=tournament_id)
                        continue

                self._process_tournament(tournament_id, details, standings, pairings, results)
                results["tournaments_processed"] += 1
                
                logger.info("Processed tournament", id=tournament_id)

            except Exception as e:
                error_msg = f"Failed to process tournament {tournament_id}: {str(e)}"
                logger.error(error_msg, exc_info=True)
                results["errors"].append(error_msg)

        self.db.conn.commit()
        return results

    def _is_processed(self, tournament_id: str) -> bool:
        cursor = self.db.conn.cursor()
        cursor.execute("SELECT COUNT(*) FROM tournaments WHERE id = ?", (tournament_id,))
        return cursor.fetchone()[0] > 0

    def _process_tournament(self, tournament_id: str, details: Dict[str, Any], standings: List[Dict[str, Any]], pairings: List[Dict[str, Any]], results: Dict[str, Any]) -> None:
        """Process a single tournament's data."""
        
        cursor = self.db.conn.cursor()

        tournament_name = details.get('name', '')
        tournament_date = details.get('date', '')
        location = None
        generation = 9
        format_val = details.get('format', '')
        official = 0

        cursor.execute("""
            INSERT OR REPLACE INTO tournaments (id, name, date, location, generation, format, official)
            VALUES (?, ?, ?, ?, ?, ?, ?)
        """, (tournament_id, tournament_name, tournament_date, location, generation, format_val, official))

        player_names_to_ids = {}

        for standing in standings:
            player_name = standing.get('name', '')
            country = standing.get('country')

            if not player_name:
                continue

            player_id = self._get_or_create_player(cursor, player_name, country)
            player_names_to_ids[player_name.lower()] = player_id

            team_id = self._get_or_create_team(cursor, player_id, tournament_id)

            placing = standing.get('placing')
            record = standing.get('record', {})
            wins = record.get('wins', 0)
            losses = record.get('losses', 0)
            ties = record.get('ties', 0)
            dropped = 1 if standing.get('drop') else 0

            cursor.execute("""
                INSERT OR REPLACE INTO tournament_standings (tournament_id, player_id, team_id, placing, wins, losses, ties, dropped)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?)
            """, (tournament_id, player_id, team_id, placing, wins, losses, ties, dropped))
            results["tournament_standings_added"] += 1

            results["players_added"] += 1
            results["teams_added"] += 1

            decklist = standing.get('decklist', [])
            for pokemon in decklist:
                pokemon_id = pokemon.get('id', pokemon.get('name', ''))
                pokemon_name = pokemon.get('name', '')
                item = pokemon.get('item')
                ability = pokemon.get('ability')
                tera_type = pokemon.get('tera')
                attacks = pokemon.get('attacks', [])

                pokemon_set_id = self._create_pokemon_set(cursor, team_id, pokemon_name, item, ability, tera_type)
                results["pokemon_sets_added"] += 1

                for move in attacks:
                    cursor.execute("""
                        INSERT INTO moves (pokemon_set_id, move_name)
                        VALUES (?, ?)
                    """, (pokemon_set_id, move))

        for pairing in pairings:
            round_number = pairing.get('round', 0)
            phase = pairing.get('phase', 1)
            table_number = pairing.get('table')

            cursor.execute("""
                INSERT INTO matches (tournament_id, round_number, table_number, phase)
                VALUES (?, ?, ?, ?)
            """, (tournament_id, round_number, table_number, phase))
            match_id = cursor.lastrowid
            results["matches_added"] += 1

            player1_name = pairing.get('player1')
            player2_name = pairing.get('player2')
            winner = pairing.get('winner')

            if player1_name:
                player1_id = player_names_to_ids.get(player1_name.lower())
                if player1_id:
                    team1_id = self._get_team_id(cursor, player1_id, tournament_id)
                    if team1_id:
                        score1 = 1 if winner == player1_name else 0
                        cursor.execute("""
                            INSERT OR REPLACE INTO match_participants (match_id, player_id, team_id, score)
                            VALUES (?, ?, ?, ?)
                        """, (match_id, player1_id, team1_id, score1))

            if player2_name:
                player2_id = player_names_to_ids.get(player2_name.lower())
                if player2_id:
                    team2_id = self._get_team_id(cursor, player2_id, tournament_id)
                    if team2_id:
                        score2 = 1 if winner == player2_name else 0
                        cursor.execute("""
                            INSERT OR REPLACE INTO match_participants (match_id, player_id, team_id, score)
                            VALUES (?, ?, ?, ?)
                        """, (match_id, player2_id, team2_id, score2))

    def _get_or_create_player(self, cursor, name: str, country: Optional[str] = None) -> int:
        cursor.execute("SELECT id FROM players WHERE name = ?", (name,))
        row = cursor.fetchone()
        
        if row:
            return row[0]
        
        cursor.execute("""
            INSERT INTO players (name, country)
            VALUES (?, ?)
        """, (name, country))
        return cursor.lastrowid

    def _get_or_create_team(self, cursor, player_id: int, tournament_id: str) -> int:
        cursor.execute("""
            SELECT id FROM teams 
            WHERE player_id = ? AND tournament_id = ?
        """, (player_id, tournament_id))
        row = cursor.fetchone()
        
        if row:
            return row[0]
        
        cursor.execute("""
            INSERT INTO teams (player_id, tournament_id)
            VALUES (?, ?)
        """, (player_id, tournament_id))
        return cursor.lastrowid

    def _get_team_id(self, cursor, player_id: int, tournament_id: str) -> Optional[int]:
        cursor.execute("""
            SELECT id FROM teams 
            WHERE player_id = ? AND tournament_id = ?
        """, (player_id, tournament_id))
        row = cursor.fetchone()
        return row[0] if row else None

    def _create_pokemon_set(self, cursor, team_id: int, species: str, item: Optional[str], ability: Optional[str], tera_type: Optional[str]) -> int:
        cursor.execute("""
            INSERT INTO pokemon_sets (team_id, species, item, ability, tera_type)
            VALUES (?, ?, ?, ?, ?)
        """, (team_id, species, item, ability, tera_type))
        return cursor.lastrowid

File: common/src/processors/processor.ts
import { logger } from '@vgc/common/logging';
import type { IDataStore } from '../database/interfaces.js';
import { validatePokemon } from './validate.js';

export interface ProcessorOptions {
  source?: string;
  tournamentIds?: string[];
  force?: boolean;
}

export class DataProcessor {
  private db: IDataStore;

  constructor(db: IDataStore) {
    this.db = db;
  }

  async processTournaments(options: ProcessorOptions = {}): Promise<Record<string, unknown>> {
    const source = options.source || 'limitless';
    const tournamentIds = options.tournamentIds;
    const force = options.force || false;

    const results: Record<string, unknown> = {
      success: true,
      tournamentsProcessed: 0,
      playersAdded: 0,
      teamsAdded: 0,
      pokemonSetsAdded: 0,
      matchesAdded: 0,
      tournamentStandingsAdded: 0,
      processedTournamentIds: [],
      errors: [],
    };

    if (source !== 'limitless') {
      results.success = false;
      (results.errors as string[]).push(`Unsupported source: ${source}`);
      return results;
    }

    const rawDataRows = await this.db.getRawData(tournamentIds);
    logger.info({ count: rawDataRows.length }, 'Found raw tournaments');

    for (const row of rawDataRows) {
      const { id: tournamentId, details: detailsJson, standings: standingsJson, pairings: pairingsJson } = row;

      try {
        if (!detailsJson || !standingsJson || !pairingsJson) {
          logger.warn({ id: tournamentId }, 'Incomplete data for tournament');
          continue;
        }

        const details = JSON.parse(detailsJson);
        const standings = JSON.parse(standingsJson);
        const pairings = JSON.parse(pairingsJson);

        if (details.game && details.game !== 'VGC') {
          logger.info({ id: tournamentId, game: details.game }, 'Skipping non-VGC tournament');
          continue;
        }

        if (!force && await this.db.isTournamentProcessed(tournamentId)) {
          logger.info({ id: tournamentId }, 'Tournament already processed, skipping');
          continue;
        }

        await this.processTournament(tournamentId, details, standings, pairings, results);
        results.tournamentsProcessed = (results.tournamentsProcessed as number) + 1;
        (results.processedTournamentIds as string[]).push(tournamentId);
        logger.info({ id: tournamentId }, 'Processed tournament');
      } catch (error) {
        const errorMsg = `Failed to process tournament ${tournamentId}: ${String(error)}`;
        logger.error({ error: String(error), tournamentId }, errorMsg);
        (results.errors as string[]).push(errorMsg);
      }
    }

    return results;
  }

  private async processTournament(
    tournamentId: string,
    details: Record<string, unknown>,
    standings: Array<Record<string, unknown>>,
    pairings: Array<Record<string, unknown>>,
    results: Record<string, unknown>,
  ): Promise<void> {
    const tournamentName = (details.name as string) || '';
    const tournamentDate = (details.date as string) || '';
    const generation = 9;
    const formatVal = (details.format as string) || '';
    const official = false;

    await this.db.upsertTournament({
      id: tournamentId,
      name: tournamentName,
      date: tournamentDate,
      location: null,
      generation,
      format: formatVal,
      official,
    });

    const playerNamesToIds: Record<string, number> = {};

    for (const standing of standings) {
      const playerName = (standing.name as string) || '';
      const country = standing.country as string | undefined;

      if (!playerName) continue;

      const alreadyKnown = playerName.toLowerCase() in playerNamesToIds;
      const playerId = await this.db.findOrCreatePlayer(playerName, country);
      playerNamesToIds[playerName.toLowerCase()] = playerId;
      if (!alreadyKnown) {
        results.playersAdded = (results.playersAdded as number) + 1;
      }

      const teamId = await this.db.findOrCreateTeam(playerId, tournamentId);
      results.teamsAdded = (results.teamsAdded as number) + 1;

      const placing = standing.placing as number | undefined;
      const record = (standing.record as Record<string, number>) || {};
      const wins = record.wins || 0;
      const losses = record.losses || 0;
      const ties = record.ties || 0;
      const dropped = !!standing.drop;

      await this.db.upsertStanding({
        tournamentId,
        playerId,
        teamId,
        placing: placing ?? null,
        wins,
        losses,
        ties,
        dropped,
      });
      results.tournamentStandingsAdded = (results.tournamentStandingsAdded as number) + 1;

      const decklist = (standing.decklist as Array<Record<string, unknown>>) || [];
      for (const pokemon of decklist) {
        const validated = validatePokemon({
          name: (pokemon.name as string) || '',
          item: pokemon.item as string | undefined,
          ability: pokemon.ability as string | undefined,
          tera: pokemon.tera as string | undefined,
          attacks: (pokemon.attacks as string[]) || [],
        });

        for (const fix of validated.fixes) {
          logger.debug({ tournamentId, playerName, fix }, 'Auto-fixed pokemon data');
        }
        for (const warning of validated.warnings) {
          logger.warn({ tournamentId, playerName, warning }, 'Unknown pokemon data');
        }

        const pokemonSetId = await this.db.insertPokemonSet({
          teamId,
          species: validated.species,
          item: validated.item,
          ability: validated.ability,
          tera_type: validated.tera_type,
          is_mega: validated.is_mega,
          invalid: validated.invalid,
        });
        results.pokemonSetsAdded = (results.pokemonSetsAdded as number) + 1;

        for (const move of validated.moves) {
          await this.db.insertMove(pokemonSetId, move);
        }
      }
    }

    for (const pairing of pairings) {
      const roundNumber = (pairing.round as number) || 0;
      const phase = (pairing.phase as number) || 1;
      const tableNumber = pairing.table as number | undefined;

      const matchId = await this.db.insertMatch({
        tournamentId,
        roundNumber,
        tableNumber: tableNumber ?? null,
        phase,
      });
      results.matchesAdded = (results.matchesAdded as number) + 1;

      const player1Name = pairing.player1 as string;
      const player2Name = pairing.player2 as string;
      const winner = pairing.winner as string;
      const winnerLower = winner?.toLowerCase() ?? '';

      if (player1Name) {
        const player1Id = playerNamesToIds[player1Name.toLowerCase()];
        if (player1Id) {
          const team1Id = await this.db.findOrCreateTeam(player1Id, tournamentId);
          await this.db.upsertMatchParticipant({
            matchId,
            playerId: player1Id,
            teamId: team1Id,
            score: winnerLower === player1Name.toLowerCase() ? 1 : 0,
          });
        }
      }

      if (player2Name) {
        const player2Id = playerNamesToIds[player2Name.toLowerCase()];
        if (player2Id) {
          const team2Id = await this.db.findOrCreateTeam(player2Id, tournamentId);
          await this.db.upsertMatchParticipant({
            matchId,
            playerId: player2Id,
            teamId: team2Id,
            score: winnerLower === player2Name.toLowerCase() ? 1 : 0,
          });
        }
      }
    }
  }
}
